ocaml/xapi/xapi_vm_lifecycle.ml

```diff
@@ -376,8 +376,11 @@ let check_operation_error ~__context ~vmr ~vmgmr ~ref ~clone_suspended_vm_enable
       match op with
       | `changing_dynamic_range ->
         let metrics = Db.VM.get_metrics ~__context ~self:ref in
+        let constr =
+          "Memory limits must satisfy: \
+          static_min <= dynamic_min = dynamic_max = static_max" in
         if nested_virt ~__context ref metrics then
-          Some (Api_errors.vm_cant_use_dyn_memory, [ref_str])
+          Some (Api_errors.memory_constraint_violation, [constr])
```

Comment:
For example, you are blocking `VM.set_memory_dynamic_range`, but not `VM.set_memory_limits`, which is capable of setting the dynamic range as well. More importantly, it looks like `Xapi_vm_memory_constraints.assert_valid_for_current_context` exists for precisely the reason you need it, and has similar checks in it already. It's best to keep this together is possible.